### DISCUSSION TITLE
Asyncio, aiogoogle, google drive... Help!

### DISCUSSION THREAD

**Discussion by ciromanfredi** [python-telegram-bot/python-telegram-bot#2344] [Q&A] [answered]
Hi, I would like some advice to solve this problem. I am developing a bot with this framework. In a function, which the user can invoke via command, an asynchronous function is called to download some parallel files from google drive. (Through asyncio and https://aiogoogle.readthedocs.io/). I would like to know how to make it work in a multi-user context.
In addition, asynchronous functions operations will be called in other functions that will be activated by commands from the various users...
how do I manage "more loop" in the whole program....<br>
async def listFilesAsync(elemIn):
&nbsp;&nbsp;&nbsp;&nbsp;async with Aiogoogle(user_creds=user_creds, client_creds=client_creds) as aiogoogle:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;async with Aiogoogle(user_creds=user_creds, client_creds=client_creds) as aiogoogle:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;drive_v3 = await aiogoogle.discover("drive", "v3")
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;....
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;....

def functionCommandStart(update:Update,context:CallbackContext):
&nbsp;&nbsp;&nbsp;&nbsp;lista=[elem1, elem2, ...]
&nbsp;&nbsp;&nbsp;&nbsp;loop = asyncio.new_event_loop() 
&nbsp;&nbsp;&nbsp;&nbsp;asyncio.set_event_loop(loop)
&nbsp;&nbsp;&nbsp;&nbsp;results = loop.run_until_complete(asyncio.gather(*[functionAsync(elem) for elem in lista]))

**Comment by zeroone2numeral2** [comment, marked as answer]
`Handler` classes offer an option to make their callback functions run asynchronously, in their own thread: [`Handler(callback, run_async=True)`](https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.handler.html#telegram.ext.Handler.run_async). The wiki has a [pretty nice section](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Performance-Optimizations#multithreading) about making the dispatcher run some handlers callbacks asynchronously.

Mind, the number of parallel updates the dispatcher will be able to run concurrently is limited to the number of workers you tell the `Dispatcher` to initialize ([docs](https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.dispatcher.html#telegram.ext.Dispatcher.workers)). You can set it while initializing your `Updater` object: `Updater(token="", workers=16)`

**Reply by ciromanfredi** [reply]
Thanks
